src/core/qgsclipper.h

```diff
@@ -326,6 +395,44 @@ inline void QgsClipper::trimPolygonToBoundary( const QPolygonF &inPts, QPolygonF
   }
 }
 
+inline void QgsClipper::trimPolygonToBoundary( const QgsLineString &inPts, QgsLineString &outPts, const QgsBox3d &rect, Boundary b, double boundaryValue )
+{
+  QgsVertexId::VertexType type;
+  QgsPoint inI1, inI2;
+
+  inPts.pointAt( inPts.numPoints() - 1, inI1, type ); // start with last point
+
+  // and compare to the first point initially.
+  for ( int i2 = 0; i2 < inPts.numPoints() ; ++i2 )
+  {
+    inPts.pointAt( i2, inI2, type );
+    // look at each edge of the polygon in turn
+    if ( inside( inI2, b, boundaryValue ) ) // end point of edge is inside boundary
+    {
+      if ( inside( inI1, b, boundaryValue ) )
+      {
+        outPts.addVertex( inI2 );
+      }
+      else
+      {
+        // edge crosses into the boundary, so trim back to the boundary, and
+        // store both ends of the new edge
+        outPts.addVertex( intersectRect( inI1, inI2, b, rect ) );
```

Comment:
addVertex is a slow call to make. It resizes the containers on every call and clears the cache each time. 

Looks like every time the trimPolygonToBoundary method is called outPts will initially be empty. I think we could improve the changes to this whole class a lot by not using QgsLineString and instead using raw QVector< double >containers for inX/outX/inY/outY/inZ/outZ. That's as close to bare metal as we can get here, and will give a good performance boost.

Reply:
arg! I was supposing it just concats the new point to the line. I'll change that!